### PR TITLE
Fix check in r_core_anal_fcn() ##anal

### LINKED ISSUE #16833: r_core_anal_fcn() doesnt analyze some symbols
### Work environment

| Questions                                            | Answers
|------------------------------------------------------|--------------------
| OS/arch/bits (mandatory)                             | any
| File format of the file you reverse (mandatory)      | any
| Architecture/bits of the file (mandatory)            | any
| r2 -v full output, **not truncated** (mandatory)     | radare2 4.5.0-git 24650 @ linux-x86-64 git.4.4.0-184-g00d933f81 commit: 00d933f81244b5a7e2496c4ded77cb7ae006db3a build: 2020-05-16__19:39:49

### Expected behavior
`af` command creates function at the given address
### Actual behavior
no new function
### Steps to reproduce the behavior 
1) `radare2 libtest.so`
2) `af @ sym.sym_lvl1`
3) `af @ sym.sym_lvl0`
4) `afl~lvl0` (output is empty) 
### Additional Logs, screenshots, source-code,  configuration dump, ...
`sym.sym_lvl1` just jumps to `sym.sym_lvl0`
image
Since bb related to `sym.sym_lvl0` was already analyzed, execution of `r_core_anal_fcn()` is stopped after this check: 
https://github.com/radareorg/radare2/blob/89cfe05d2d5a3899d4dd846f6e853f9a12769ae1/libr/core/canal.c#L1934-L1939
Drag and drop zip archives containing the Additional info here, don't use external services or link.
[libtest.zip](https://github.com/radareorg/radare2/files/4639342/libtest.zip)

Blocker of #7310.

### PR DESCRIPTION
**Pull request by pelijah**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md#code-style)
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [radare2 book](https://github.com/radareorg/radare2book) with the relevant information (if needed)

**Detailed description**

...

**Test plan**

...

**Closing issues**

Fixes #16833
...

### CODE REVIEW

**Comment by pelijah** [comment]
Nice... "fix" reveals another issue.

File: libr/core/canal.c [hunk 1]
```diff
@@ -1926,12 +1926,21 @@ R_API int r_core_anal_esil_fcn(RCore *core, ut64 at, ut64 from, int reftype, int
 	return 0;
 }
 
+static int find_sym_flag(RAnalFunction *fcn, RFlagItem *f) {
+	return f->space && !strcmp (f->space->name, R_FLAGS_FS_SYMBOLS)? 0: 1;
+}
+
+static bool is_invalid_addr(RCore *core, ut64 addr) {
```
Comment (radare) [inline libr/core/canal.c:1933]: just this function just call r_io_is_valid api to determine if its mapped or not? if theres a function or a flag doesnt matters much imho. also i think this slowdowns the analysis

**Comment by XVilka** [comment]
Very red.

File: libr/core/canal.c (hunk 1 above)
Comment (pelijah) [reply libr/core/canal.c:1933]: IDK who wrote this check and why it prints error "Invalid address...". I just extended it, moved it to a function and named it `is_invalid_addr`.

**Comment by pelijah** [comment]
> 
> 
> Very red.

@XVilka Yeah, because handling of delayed slots in r2 is a crap.

**Comment by pelijah** [comment]
I want to find better binary for test.

File: libr/core/canal.c (hunk 1 above)
Comment (pelijah) [reply libr/core/canal.c:1933]: I can rename it to avoid confusions. Maybe `skip_addr`?

Comment (pelijah) [reply libr/core/canal.c:1933]: I don't see performance loss on the results of performed tests. The check is involved only when `r_core_anal_fcn()` is called inside `aa`, `af` and `afu`.

**Comment by ret2libc** [comment]
@pelijah is this ready for review?

**Comment by pelijah** [comment]
Yep

**Comment by pelijah** [comment]
Well it would be better to find other solution which will solve #16979 too.

File: libr/core/canal.c (hunk 1 above)
Comment (trufae) [reply libr/core/canal.c:1933]: what this function is doing is not checking if the address is valid. this is what RIO.is_valid_addr() do.

In this case this is chcking if there's a function or a flag in there. Fix the function name to clarify this and LGTM

**Comment by trufae** [comment]
still red, but thanks for addressing the comments

File: libr/core/canal.c [hunk 2]
```diff
@@ -1926,12 +1926,28 @@ R_API int r_core_anal_esil_fcn(RCore *core, ut64 at, ut64 from, int reftype, int
 	return 0;
 }
 
+static int find_sym_flag(RAnalFunction *fcn, RFlagItem *f) {
```
Comment (ret2libc) [inline libr/core/canal.c:1929]: the right signature is `(void *, void *)`

File: libr/core/canal.c [hunk 3]
```diff
@@ -1926,12 +1926,28 @@ R_API int r_core_anal_esil_fcn(RCore *core, ut64 at, ut64 from, int reftype, int
...
+
+static bool is_skippable_addr(RCore *core, ut64 addr) {
+	RAnalFunction *fcn = r_anal_get_fcn_in (core->anal, addr, 0);
+	if (!fcn) {
+		return false;
+	}
+	if (fcn->addr == addr) {
+		return true;
+	}
+	RList *flags = r_flag_get_list (core->flags, addr);
```
Comment (ret2libc) [inline libr/core/canal.c:1941]: `const RList *`

**Comment by XVilka** [comment]
@pelijah could you please address the feedback, so we can merge it before the release? It looks important.

**Comment by pelijah** [comment]
To the first commit?

File: libr/core/canal.c [hunk 4]
```diff
@@ -1926,12 +1926,29 @@ R_API int r_core_anal_esil_fcn(RCore *core, ut64 at, ut64 from, int reftype, int
...
+
+static bool is_skippable_addr(RCore *core, ut64 addr) {
+	RAnalFunction *fcn = r_anal_get_fcn_in (core->anal, addr, 0);
+	if (!fcn) {
+		return false;
+	}
+	if (fcn->addr == addr) {
+		return true;
+	}
+	const RList *flags = r_flag_get_list (core->flags, addr);
```
Comment (trufae) [inline libr/core/canal.c:1942]: what about using rbin instead of flags to determine if there's a symbol? cc @ret2libc @pelijah ?

**Comment by trufae** [comment]
to the pr title, i just added it. im not 100% for checking flags to find symbols, but at least is faster than iterating over rbin symbol lists.

File: libr/core/canal.c (hunk 4 above)
Comment (ret2libc) [reply libr/core/canal.c:1942]: I'm not sure how the change would affect the analysis, it should be tried.

Comment (XVilka) [reply libr/core/canal.c:1942]: I suggest we merge this as is, but consider this change in a new PR.

**Pull request merged**